sdv/cag/base.py

```diff
@@ -0,0 +1,166 @@
+"""Base Multi-Table Pattern."""
+
+import pandas as pd
+
+from sdv.errors import NotFittedError
+
+
+class BasePattern:
+    """Base CAG Pattern Class."""
+
+    def __init__(self):
+        self.metadata = None
+        self._fitted = False
+        self._single_table = False
+
+    def _get_single_table_name(self, metadata):
+        if not hasattr(self, 'table_name'):
+            raise ValueError('No ``table_name`` attribute has been set.')
+
+        return metadata._get_single_table_name() if self.table_name is None else self.table_name
+
+    def _validate_pattern_with_metadata(self, metadata):
+        raise NotImplementedError()
+
+    def _validate_pattern_with_data(self, data, metadata):
+        raise NotImplementedError()
+
+    def validate(self, data=None, metadata=None):
+        """Validate the data/metadata meets the pattern requirements.
+
+        Args:
+            data (dict[str, pd.DataFrame], optional)
+                The data dictionary. If `None`, ``validate`` will skip data validation.
+            metadata (sdv.Metadata, optional)
+                The input metadata. If `None`, pattern must have been fitted and ``validate``
+                will use the metadata saved during fitting.
+        """
+        if metadata is None:
+            if self.metadata is None:
+                raise NotFittedError('Pattern must be fit before validating without metadata.')
+
+            metadata = self.metadata
+
+        self._validate_pattern_with_metadata(metadata)
+        if data is not None:
+            if self._single_table:
+                data = {self._table_name: data}
```

Comment:
There are `self._table_name` and `self.table_name`, it seems; should they be the same?

Reply:
Not necessarily, the base class learns `self._table_name` during fit if it's a single table. `self.table_name` stores whatever the actual `table_name` parameter was. If the input `self.table_name` is `None`, then we don't want `self._table_name` to break the pattern on a refit in case the table name has changed.